packages/iov-core/src/jsrpc.ts

```diff
@@ -0,0 +1,242 @@
+import {
+  jsonRpcCodeInternalError,
+  jsonRpcCodeInvalidParams,
+  jsonRpcCodeInvalidRequest,
+  jsonRpcCodeMethodNotFound,
+  jsonRpcCodeParseError,
+  jsonRpcCodeServerErrorDefault,
+  SimpleMessagingConnection,
+} from "@iov/jsonrpc";
+import { firstEvent } from "@iov/stream";
+
+// An RPC interface very similar to JSON-RPC but supporting
+// all the data types from the structured clone algorithm
+// https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#Supported_types
+
+/** A single structured clone algorithm compatible value */
+export declare type JsRpcCompatibleValue =
+  | string
+  | number
+  | boolean
+  | null
+  | Uint8Array
+  | JsRpcCompatibleArray
+  | JsRpcCompatibleDictionary;
+
+/** An array of JsRpcCompatibleValue */
+export interface JsRpcCompatibleArray extends ReadonlyArray<JsRpcCompatibleValue> {}
+
+/** A string to JsRpcCompatibleValue dictionary. */
+export interface JsRpcCompatibleDictionary {
+  readonly [key: string]: JsRpcCompatibleValue;
+}
+
+export function isJsRpcCompatibleValue(value: unknown): value is JsRpcCompatibleValue {
+  return (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean" ||
+    value === null ||
+    value instanceof Uint8Array ||
+    isJsRpcCompatibleArray(value) ||
+    isJsRpcCompatibleDictionary(value)
+  );
+}
+
+export function isJsRpcCompatibleArray(value: unknown): value is JsRpcCompatibleArray {
+  if (!Array.isArray(value)) {
+    return false;
+  }
+  return value.every(isJsRpcCompatibleValue);
+}
+
+export function isJsRpcCompatibleDictionary(data: unknown): data is JsRpcCompatibleDictionary {
+  if (typeof data !== "object" || data === null) {
+    // data must be a non-null object
```

Comment:
Is it ever possible that `typeof data === "object" && data === null` returns true?
Seems like a needless second check unless you can show me a case where needed (maybe a link)

Reply:
The thing is: in JavaScript, `null` is an object:

```
> typeof null
'object'
```

This is what I wanted to express with that comment.

Reply:
Wow. I didn't realize that... Maybe update the comment to read:

`// the thing is: in JavaScript, null is an object`